File: src/main/java/dev/lampirg/consultationappointment/service/student/SimpleAppointmentMaker.java
package dev.lampirg.consultationappointment.service.student;

import dev.lampirg.consultationappointment.data.appointment.Appointment;
import dev.lampirg.consultationappointment.data.appointment.AppointmentRepository;
import dev.lampirg.consultationappointment.data.student.Student;
import dev.lampirg.consultationappointment.data.student.StudentRepository;
import dev.lampirg.consultationappointment.data.teacher.DatePeriod;
import dev.lampirg.consultationappointment.data.teacher.Teacher;
import org.springframework.stereotype.Service;

import java.time.Duration;
import java.time.LocalDate;

@Service
public class SimpleAppointmentMaker implements AppointmentMaker {

    public final static Duration INTERVAL = Duration.ofMinutes(15);

    private AppointmentRepository appointmentRepository;
    private StudentRepository studentRepository;

    public SimpleAppointmentMaker(AppointmentRepository appointmentRepository, StudentRepository studentRepository) {
        this.appointmentRepository = appointmentRepository;
        this.studentRepository = studentRepository;
    }

    @Override
    public void makeAppointment(Teacher teacher, Student student, DatePeriod datePeriod) {
        if (!isAvailable(teacher, student, datePeriod))
            throw new IllegalArgumentException("Not enough time for new appointment");
        Appointment appointment = new Appointment();
        appointment.setStartTime(datePeriod.getStartTime());
        appointment.setTeacher(teacher);
        appointment.setStudent(student);
        appointment.setAppointmentPeriod(datePeriod);
        datePeriod.setUnoccupiedTime(
                datePeriod.getUnoccupiedTime().minus(INTERVAL)
        );
        appointment = appointmentRepository.save(appointment);
        studentRepository.findById(appointment.getStudent().getId()).get().getAppointments().add(appointment);
    }

    @Override
    public void deleteAppointmentById(Long id) {
        Appointment appointment = appointmentRepository.findById(id).orElseThrow();
        appointment.getAppointmentPeriod().setUnoccupiedTime(
                appointment.getAppointmentPeriod().getUnoccupiedTime().plus(INTERVAL)
        );
        appointmentRepository.save(appointment);
        appointmentRepository.delete(appointment);
        studentRepository.findById(appointment.getStudent().getId()).get().getAppointments().remove(appointment);
    }

    @Override
    public boolean isAvailable(DatePeriod datePeriod) {
        return !datePeriod.getUnoccupiedTime().minus(INTERVAL).isNegative() &&
                (datePeriod.getStartTime().toLocalDate().isAfter(LocalDate.now()) ||
                        !datePeriod.getAppointments().isEmpty());
    }

    @Override
    public boolean isAvailable(Teacher teacher, Student student, DatePeriod datePeriod) {
        if (!isAvailable(datePeriod))
            return false;
        return student.getAppointments().stream()
                .map(Appointment::getAppointmentPeriod)
                .noneMatch(appointmentPeriod -> appointmentPeriod.equals(datePeriod));
    }
}

File: src/main/kotlin/dev/lampirg/consultationappointment/service/student/SimpleAppointmentMaker.kt
package dev.lampirg.consultationappointment.service.student

import dev.lampirg.consultationappointment.data.appointment.Appointment
import dev.lampirg.consultationappointment.data.appointment.AppointmentRepository
import dev.lampirg.consultationappointment.data.student.Student
import dev.lampirg.consultationappointment.data.student.StudentRepository
import dev.lampirg.consultationappointment.data.teacher.DatePeriod
import dev.lampirg.consultationappointment.data.teacher.Teacher
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Duration
import java.time.LocalDate

@Service
class SimpleAppointmentMaker @Autowired constructor(
    private val appointmentRepository: AppointmentRepository,
    private val studentRepository: StudentRepository
) :
    AppointmentMaker {
    @Transactional
    override fun makeAppointment(teacher: Teacher, student: Student, datePeriod: DatePeriod) {
        require(isAvailable(teacher, student, datePeriod)) { "Not enough time for new appointment" }
        var appointment = Appointment(teacher, student, datePeriod, datePeriod.startTime)
        datePeriod.unoccupiedTime = datePeriod.unoccupiedTime.minus(INTERVAL)
        appointment = appointmentRepository.save(appointment)
        studentRepository.findById(appointment.student.id!!).orElseThrow().appointments.add(appointment)
    }

    @Transactional
    override fun deleteAppointmentById(id: Long) {
        val appointment = appointmentRepository.findById(id).orElseThrow()!!
        appointment.appointmentPeriod.unoccupiedTime = appointment.appointmentPeriod.unoccupiedTime.plus(INTERVAL)
        appointmentRepository.save(appointment)
        appointmentRepository.delete(appointment)
        studentRepository.findById(appointment.student.id!!).orElseThrow().appointments.remove(appointment)
    }

    override fun isAvailable(datePeriod: DatePeriod): Boolean {
        return !datePeriod.unoccupiedTime.minus(INTERVAL).isNegative &&
                (datePeriod.startTime.toLocalDate().isAfter(LocalDate.now()) ||
                        datePeriod.appointments.isNotEmpty())
    }

    override fun isAvailable(teacher: Teacher, student: Student, datePeriod: DatePeriod): Boolean {
        return if (!isAvailable(datePeriod)) false else student.appointments.stream()
            .map(Appointment::appointmentPeriod)
            .noneMatch { appointmentPeriod: DatePeriod -> appointmentPeriod.equals(datePeriod) }
    }

    companion object {
        val INTERVAL: Duration = Duration.ofMinutes(15)
    }
}